src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs

```diff
@@ -0,0 +1,159 @@
+
+            ModifiedPages = new HashSet<long>();
+            FreedPages = new HashSet<long>();
+
+            Tree.PageModified += page => ModifiedPages.Add(page);
+            Tree.PageFreed += page => FreedPages.Add(page);
+        }
+
+        public void Delete(long id)
+        {
+            var entryId = id;
+
+            switch (Type)
+            {
+                case MapResultsStorageType.Tree:
+                    Tree.Delete(Slice.External(_indexContext.Allocator, (byte*)&entryId, sizeof(long)));
+                    break;
+                case MapResultsStorageType.Nested:
+                    var section = GetNestedResultsSection();
+
+                    section.Delete(id);
+
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(Type.ToString());
+            }
+            
+            _mapReduceContext.EntryDeleted(id);
+        }
+
+        public void Add(long id, BlittableJsonReaderObject result, bool isUpdate)
+        {
+            switch (Type)
+            {
+                case MapResultsStorageType.Tree:
+                    var pos = Tree.DirectAdd(Slice.External(_indexContext.Allocator, (byte*)&id, sizeof(long)), result.Size);
+                    result.CopyTo(pos);
+                    break;
+                case MapResultsStorageType.Nested:
+                    var section = GetNestedResultsSection();
+
+                    section.Add(id, result, isUpdate);
+
+                    if (_mapReduceContext.MapEntries.ShouldGoToOverflowPage(_nestedSection.Size))
+                    {
+                        // would result in an overflow, that would be a space waste anyway, let's move to tree mode
+                        MoveExistingResultsToTree(section);
+
+                        _nestedSection.Dispose();
+                        _nestedSection = null;
+
+                        _mapReduceContext.MapEntries.Delete(_nestedValueKey);
+                    }
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(Type.ToString());
+            }
+        }
+
+        private void MoveExistingResultsToTree(NestedMapResultsSection section)
+        {
+            Type = MapResultsStorageType.Tree;
+
+            var byteType = (byte)Type;
+
+            _mapReduceContext.ResultsStoreTypes.Add((long)_reduceKeyHash, Slice.External(_indexContext.Allocator, &byteType, sizeof(byte)));
+
+            InitializeTree(create: true);
+
+            foreach (var mapResult in section.GetResults())
+            {
+                Add(mapResult.Key, mapResult.Value, false);
+            }
+        }
+
+        private NestedMapResultsSection _nestedSection;
+
+        public NestedMapResultsSection GetNestedResultsSection()
+        {
+            if (_nestedSection != null)
+                return _nestedSection;
+
+            var read = _mapReduceContext.MapEntries.Read(_nestedValueKey);
```

Comment:
`BasicAutoMapReduceIndexing.CanDelete` failed after this change (in addition to the arena issues). The problem was that we leave behind zero length values on delete. Shouldn't be cleaning it completely?